### ISSUE TITLE
[NEED ASSIST] Cryptor.swift compilation error on Cocoapods

### ISSUE THREAD

**Issue by Fletcher-FL** [Adyen/adyen-ios#561]
**Describe the challenge**
After updating to Adyen 4.0.0 using Cocoapods I'm getting the following error in `Cryptor.swift` line 78:
`Value of optional type 'Data?' must be unwrapped to refer to member 'base64EncodedString' of wrapped base type 'Data'`

Xcode 12.4, Swift 5.

**Comment by descorp** [comment]
Hey @Fletcher-FL 

And you can't build the project because of this?
This is odd. From where I can see it, [the line 78](https://github.com/Adyen/adyen-ios/blob/master/AdyenEncryption/Cryptor/Cryptor.swift#L78) all possible `Data` instances are non-optional. 

I can only suggest to cleanup the project and do `pod update` 🤔
If error remains - could you share the code you have in your pod?

**Comment by Fletcher-FL** [comment]
@descorp
I tired `pod update` the issue still remains. I checked and had a colleague also update to 4.0.0 and they see the same error as well. Did you want to see the `Cryptor.swift` file or a different file?

**Comment by descorp** [comment]
Yes, the `Cryptor.swift` file, please.

**Comment by Fletcher-FL** [comment]
@descorp I have attached the zipped `Cryptor.swift` file to this message in addition to the pasted code.

[Cryptor.swift.zip](https://github.com/Adyen/adyen-ios/files/6818300/Cryptor.swift.zip)

`//
// Copyright (c) 2021 Adyen N.V.
//
// This file is open source and available under the MIT license. See the LICENSE file for more info.
//

import CommonCrypto
import Foundation

/// :nodoc:
internal struct Cryptor {

    private static let msgSeparator = "$"

    /// :nodoc:
    internal let aes: AES

    private var prefix: String {
        "adyenio_0_\(aes.rawValue)_25" + Cryptor.msgSeparator
    }

    /// Initiate cryptor with selected AES alghoritm
    /// :nodoc:
    internal init(aes: AES) {
        self.aes = aes
    }
    
    /**
     * Encrypts the data with AES-CBC using
     * generated AES256 session key and IV (12)
     * Encrypts the session key with RSA using
     * public key (using Keychain)
     *
     *
     * - Parameters:
     *   - data: Data to be encrypted.
     *   - keyInHex: Public key in Hex with format "Exponent|Modulus"
     *
     * - Returns: Fully composed message in format:
     *   - a prefix
     *   - a separator
     *   - RSA encrypted AES key, base64 encoded
     *   - a separator
     *   - a Payload of iv and cipherText, base64 encoded
     *
     */
    internal func encrypt(data: Data, publicKey keyInHex: String) throws -> String {
        
        //
        var key = [UInt8](repeating: 0, count: kCCKeySizeAES256)
        guard SecRandomCopyBytes(kSecRandomDefault, key.count, &key) == noErr else {
            throw Error.randomGenerationError
        }
        
        // generate a initialisation vector
        var initVector = [UInt8](repeating: 0, count: aes.ivLength)
        guard SecRandomCopyBytes(kSecRandomDefault, initVector.count, &initVector) == noErr else {
            throw Error.randomGenerationError
        }
        
        guard
            case let encryptionResult = self.aesEncrypt(data: data, with: Data(key), initVector: Data(initVector)),
            let cipherText = encryptionResult.cipher
        else { throw Error.aesEncryptionError }
        
        // format of the fully composed message:
        // - a prefix
        // - a separator
        // - RSA encrypted AES key, base64 encoded
        // - a separator
        // - a Payload of iv and cipherText, base64 encoded
        var payload = Data()
        payload.append(contentsOf: initVector)
        payload.append(cipherText)
        payload.append(encryptionResult.atag ?? Data())
        
        if let encryptedKey = try? self.rsaEncrypt(data: Data(key), with: keyInHex) {
            return prefix + encryptedKey.base64EncodedString() + Cryptor.msgSeparator + payload.base64EncodedString()
        }
        
        throw Error.rsaEncryptionError
    }
    
    private func aesEncrypt(data: Data, with key: Data, initVector: Data) -> EncryptionResult {
        aes.encrypt(data: data as NSData, withKey: key as NSData, initVector: initVector as NSData)
    }
    
    private func rsaEncrypt(data: Data, with keyInHex: String) throws -> Data? {
        let tokens = keyInHex.components(separatedBy: "|")
        guard tokens.count == 2 else { return nil }
        
        return try RSA.encrypt(data: data, exponent: tokens[0], modulus: tokens[1])
    }

    internal enum Error: Swift.Error, LocalizedError {
        case randomGenerationError
        case aesEncryptionError
        case rsaEncryptionError

        public var errorDescription: String? {
            switch self {
            case .randomGenerationError:
                return "Error occured while genrating an array of cryptographically secure random bytes."
            case .aesEncryptionError:
                return "Error occured while encryptng data."
            case .rsaEncryptionError:
                return "Error occured while encryptng session key."
            }
        }
    }
}
`

**Comment by descorp** [comment]
Hey @Fletcher-FL 

I have tested on `clean VM` locally (so it is a bit more sophisticated than "It works on my machine" 😆) and I wasn't able to reproduce the error.

`Cryptor.swift` that I am getting from Cocoapods and the one you have send above are identical, so I can only assume Xcode indexing or caching issue. 
Have you tried cleaning up Derived data?

```bash
rm -rv ~/Library/Developer/Xcode/DerivedData
```
---

In general, if you have a closer look at the line, both  `payload` and `encryptedKey` are non-optional:

```swidft
        var payload = Data()
        ...
        
        if let encryptedKey = try? self.rsaEncrypt(data: Data(key), with: keyInHex) {
            return prefix + encryptedKey.base64EncodedString() + Cryptor.msgSeparator + payload.base64EncodedString()
        }

        throw Error.rsaEncryptionError
 ```

So I am doubtful that the issue is on `Cryptor.swif`.. 
You don't have any local modifications or extensions that might interfere?

**Comment by Fletcher-FL** [comment]
I tired clearing out derived data, also no luck.

 However, I did edit the file to force unwrap `encrptedKey` just to see what would happen. The error went away and I got a slew of new errors that correspond to the breaking changes from 3.8.5 -> 4.0.0. This could be a symptom of our code not being in compliance with the SDK. Unfortunately, at this time we don't get the required `clientKey` from our backend to upgrade to 4.0.0 so I'm not able to test this hypothesis and make the upgrade to 4.0.0.

Closing this issue until we can upgrade to 4.0.0.
